### PR TITLE
test: harden Rust engine boundaries

### PR DESCRIPTION
**Pull request by anmorgunov**
## why

The native engine owns artifact parsing, provider adaptation, route identity, parallel scoring, analysis, and the standalone CLI. Its tests need to make those risky changes cheap, not merely increase coverage or prove that random input does not panic.

The suite follows the repository's `.practices` testing philosophy: pure invariants get narrow unit/property tests, ownership boundaries get contract tests, cross-implementation behavior gets differential tests, and the standalone user path gets small subprocess integration tests.

## what changed

- Adds shrinking `proptest` generators in a dedicated `retrocast-core/tests/properties.rs` target:
  - route paths and scalar wire invariants;
  - candidate outcome invariants;
  - generated route-tree traversal, depth, identity, serialization, and annotation/sibling-order metamorphisms;
  - stock and route-depth constraint truth tables;
  - bootstrap bounds, seed stability, constant-sample intervals, pairwise antisymmetry, and ranking probability laws.
- Differentially compares generated multi-target streaming ingestion with owned in-memory ingestion across 1–12 workers, target IDs, SMILES aliases, ID precedence, failures, and irrelevant providers.
- Requires adapter candidate boundaries to be worker-deterministic, respect candidate limits, remain serializable, and preserve rank/outcome invariants for bounded generated JSON.
- Requires full ingest → score → analyze semantics to survive worker-count changes, candidate storage reordering, duplicate metric options, and appended late failures.
- Keeps malformed bytes as an explicitly secondary panic-safety backstop, alongside surgical regressions for corrupt/truncated gzip, JSONL row errors, deep JSON, unsafe paths, CLI side effects, and partial argument sets.
- Gives Rust its own bounded CI workflow for formatting, Clippy, all workspace targets, and 512 generated cases per Rust property. Python workflows no longer own Rust lint/test gates.
- Fixes a bug found by the adversarial audit: standalone Rust analysis accepted `n_boot=0` and could panic while computing percentiles. It now returns a typed error before doing work; the CLI regression requires a clean non-zero exit with no output artifact.

## risk

The new production behavior is deliberately narrow: zero bootstrap resamples now return `InvalidBootstrapResamples(0)` instead of reaching a panic. Positive resample counts and all existing analysis output are unchanged.

Property generators are recursively and cardinally bounded, and the Rust job has a 15-minute timeout. Failing `proptest` cases report a replayable seed and shrink to a minimal counterexample.

## verification

- `PROPTEST_CASES=512 cargo test --manifest-path packages/retrocast-rs/Cargo.toml -p retrocast-core --locked` — 80 passed
- `cargo test --manifest-path packages/retrocast-rs/Cargo.toml --workspace --all-targets --locked` — 87 passed
- `cargo clippy --manifest-path packages/retrocast-rs/Cargo.toml --workspace --all-targets --locked -- -D warnings`
- `cargo fmt --manifest-path packages/retrocast-rs/Cargo.toml --all -- --check`
- `git diff --check`

## Summary by CodeRabbit

- **Bug Fixes**
  - Added explicit validation errors for zero bootstrap resamples and invalid worker counts.
  - Improved robustness against malformed, corrupted (including gzip), truncated, and deeply nested inputs.
  - Hardened CLI against unsafe path selectors, path traversal directives, and incomplete/invalid file-mode options, preventing unintended output creation.
- **Reliability**
  - Strengthened deterministic results across worker counts and consistent adapter listings.
- **Testing**
  - Added extensive adversarial and property-based coverage across core workflows and CLI/contract behavior.
- **CI / Quality**
  - Updated CI workflows to improve coverage reporting, enforce lint/format gates, and streamline test execution.

<h3>Greptile Summary</h3>

This PR hardens the Rust engine's test boundaries by adding property-based, differential, contract, and adversarial tests for artifact I/O, route topology, adapter boundaries, bootstrap statistics, and the standalone CLI. It also delivers one narrow production fix: `n_boot=0` now returns `InvalidBootstrapResamples(0)` instead of reaching a panic in the percentile computation. Rust lint, formatting, and testing are promoted to a dedicated CI workflow.

- **Production fix**: `validate_analysis_options` is called at the entry points of both `analyze` and `summarize_target_results` before any parallelism, so the guard is correctly placed and well-exercised by both unit and CLI regression tests.
- **Property generators** in `properties.rs` are recursively and cardinally bounded (depth 4, max 32 nodes, strings 0–63 chars), avoiding runaway generation while still covering non-trivial trees.
- **Differential test** in `adversarial.rs` compares streamed file ingestion against in-memory ingestion across 1–12 workers for three molecules with generated aliases, key modes, and failure permutations, providing strong determinism coverage.
- **CI workflow** (`rust.yml`) separates Rust gates from the Python workflow and runs property tests at 512 cases after the coverage pass.

<h3>Confidence Score: 5/5</h3>

Safe to merge. The only production change is the narrow `n_boot=0` guard in `validate_analysis_options`, which is correctly placed before any parallel work and is covered by two independent test paths.

All changed production code is minimal and directly exercised by the new tests. The `validate_analysis_options` guard is invoked at both public entry points and confirmed clean at all positive resample counts. The rest of the diff is test infrastructure and CI plumbing — no existing behaviour is removed or altered.

`.github/workflows/rust.yml` has a minor workflow design gap where the coverage report comment is not posted when coverage drops below the minimum threshold.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/retrocast-rs/crates/retrocast-core/src/analyze.rs | Adds `validate_analysis_options` guard before any parallel work; correctly covers both public entry points and is tested by adversarial and CLI regression tests. |
| packages/retrocast-rs/crates/retrocast-core/src/error.rs | Adds `InvalidBootstrapResamples(usize)` variant with a clear error message; all other variants unchanged. |
| packages/retrocast-rs/crates/retrocast-core/tests/properties.rs | Well-bounded proptest generators covering route topology, path round-trips, scalar wire invariants, candidate outcome invariants, bootstrap seed stability, antisymmetry, and ranking probability laws. |
| packages/retrocast-rs/crates/retrocast-core/tests/adversarial.rs | Covers malformed bytes (panic-safety backstop), compression determinism, streamed vs. owned ingest differential, adapter worker determinism, zero-workers rejection, and zero-bootstrap-resamples rejection. |
| packages/retrocast-rs/crates/retrocast-core/tests/workflow_contracts.rs | Integration-level contracts covering curation excision/pruning, provenance verification, standalone pipeline output, training release builders, and dataset download cache repair; all use real in-process calls. |
| packages/retrocast-rs/crates/retrocast-core/tests/evaluation_contracts.rs | Tests analysis denominator semantics (failed targets stay in denominators), scoring constraint combination, and target overrides with prefix/exact match modes. |
| packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs | CLI adversarial tests covering path-traversal rejection (with exact diagnostic assertion), malformed inputs, invalid execution options, partial file-mode argument rejection, and adapter listing completeness. |
| packages/retrocast-rs/crates/retrocast-cli/tests/cli_contracts.rs | Full pipeline-to-verify round-trip with tampering detection, and get-data dry-run dependency expansion; both use the compiled binary via CARGO_BIN_EXE_retrocast. |
| .github/workflows/rust.yml | New Rust CI workflow with pinned actions (SHAs + version comments), coverage threshold, and a second property-only run at 512 cases; the coverage-comment job has no `always()` guard so comments are suppressed when coverage drops below threshold. |
| .github/workflows/tests.yml | Removes the Rust test step from the Python workflow; renames the job for clarity. |
| .github/workflows/typing.yml | Removes Rust fmt/clippy from the typing workflow since they now live in rust.yml. |
| packages/retrocast-rs/crates/retrocast-core/tests/dataset_contracts.rs | Contracts for training data dry-run resolution and hosted benchmark target expansion; both use fake checksums (hash of path strings) which is correct for dry-run scenarios. |

<sub>Reviews (7): Last reviewed commit: ["test: cover Rust evaluation and delivery..."](https://github.com/ischemist/project-procrustes/commit/e3aec11dc41c1305369a2e98f8103b66aa1ccb02) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44700812)</sub>

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+          echo "RDKIT_ROOT=$CONDA_PREFIX" >> "$GITHUB_ENV"
+          echo "BOOST_ROOT=$CONDA_PREFIX" >> "$GITHUB_ENV"
+          echo "LD_LIBRARY_PATH=$CONDA_PREFIX/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}" >> "$GITHUB_ENV"
+
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          components: clippy, rustfmt
+
+      - name: Cache Cargo build
+        uses: actions/cache@v5
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply .github/workflows/rust.yml]: Addressed in 42a1ed1. The workflow now grants only `contents: read`, disables checkout credential persistence, and pins checkout v6, setup-micromamba v3, rust-toolchain stable, and cache v5 to their currently resolved immutable commit SHAs.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+on:
+  push:
+    branches:
+      - main
+      - master
+  pull_request:
+    branches:
+      - main
+      - master
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply .github/workflows/rust.yml:11]: Addressed in 42a1ed1. Added a workflow/ref-scoped concurrency group with `cancel-in-progress: true`, so superseded Rust runs are cancelled.

File: packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        raw.to_str().unwrap(),
+        "--benchmark",
+        directory.path().join("missing-task.json").to_str().unwrap(),
+        "--stock",
+        directory.path().join("missing-stock.csv").to_str().unwrap(),
+        "--output-dir",
+        output.to_str().unwrap(),
+    ]);
+    assert!(!output.exists());
+}
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs:146]: Addressed in 42a1ed1. The test now asserts the exact `unsafe raw_results_filename directive` diagnostic, so a missing benchmark or stock error cannot satisfy the test.

File: packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+#[test]
+fn partial_file_mode_arguments_are_rejected_without_side_effects() {
+    let directory = tempdir().unwrap();
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs:150]: Addressed in 42a1ed1. The missing external inputs remain intentional because the test claims traversal is rejected before they are read; it now asserts the exact unsafe `raw_results_filename` diagnostic, which proves that ordering and prevents an unrelated missing-file failure from passing.

File: packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        .filter(|line| !line.contains("deprecated alias"))
+        .collect::<Vec<_>>();
+    assert_eq!(canonical.len(), 13);
+    let mut unique = canonical.clone();
+    unique.sort_unstable();
+    unique.dedup();
+    assert_eq!(unique.len(), canonical.len());
+}
+
+fn assert_clean_failure(arguments: &[&str]) {
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply packages/retrocast-rs/crates/retrocast-cli/tests/adversarial_cli.rs:189]: Addressed in 42a1ed1 by removing the magic count. The canonical CLI output is now compared directly with `BUILT_IN_ADAPTERS`, so additions, removals, ordering changes, or omissions cannot silently drift.

File: .github/workflows/rust.yml [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+permissions:
+  contents: read
+  pull-requests: write
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply .github/workflows/rust.yml]: Addressed in b3f0be6. The validation job now retains only `contents: read`; coverage commenting runs in a separate `needs: test` job with job-scoped `pull-requests: write`, no checkout, and guards that skip fork and Dependabot PRs.

File: packages/retrocast-rs/crates/retrocast-core/tests/adversarial.rs [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+        "prefix",
+        None,
+        1,
+    )
+    .unwrap();
+
+    let error = analyze(&evaluation, &[1], &[1], 0, 42, 1).unwrap_err();
+
+    assert!(matches!(error, EngineError::InvalidBootstrapResamples(0)));
+}
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply packages/retrocast-rs/crates/retrocast-core/tests/adversarial.rs:374]: Addressed in b3f0be6. The zero-bootstrap regression now calls both public entry points and requires `summarize_target_results` to return `EngineError::InvalidBootstrapResamples(0)` as well.
